rescript/get_midori2.py

```diff
@@ -0,0 +1,117 @@
+# ----------------------------------------------------------------------------
+# Copyright (c) 2019-2025, QIIME 2 development team.
+#
+# Distributed under the terms of the Modified BSD License.
+#
+# The full license is in the file LICENSE, distributed with this software.
+# ----------------------------------------------------------------------------
+
+import os
+import tempfile
+import shutil
+import gzip
+from urllib.request import urlretrieve
+from pathlib import Path
+import pandas as pd
+from q2_types.feature_data import (TaxonomyFormat, DNAFASTAFormat,
+                                   DNAIterator)
+
+MITO_GENE_LIST = ['A6', 'A8', 'CO1', 'CO2', 'CO3', 'Cytb',
+                  'ND1', 'ND2', 'ND3', 'ND4L', 'ND4',
+                  'ND5', 'ND6', 'lrRNA', 'srRNA', 'all']
+
+
+def _assemble_midori2_urls(mito_gene,
+                           version='GenBank265_2025-03-08',
+                           ref_seq_type='uniq',
+                           unspecified_species=False,
+                           ):
+    print('Building URLs ...')
+    if unspecified_species:
+        spec = 'QIIME_sp'
+        sp = 'SP_'
+    else:
+        spec = 'QIIME'
+        sp = ''
+
+    vernum = version.strip('GenBank').split('_')[0]
+
+    base_url = ('https://www.reference-midori.info/download/Databases/'
+                '{full_version}/{spec}/{ref_seq_type}/MIDORI2_'
+                '{ref_seq_type_u}_NUC_'
+                '{sp}GB{vernum}_{mito_gene}_QIIME.{file_type}.gz')
+
+    base_format_dict = {'full_version': version,
+                        'spec': spec,
+                        'ref_seq_type': ref_seq_type,
+                        'ref_seq_type_u': ref_seq_type.upper(),
+                        'sp': sp,
+                        'vernum': vernum,
+                        'mito_gene': mito_gene}
+
+    fasta_url = base_url.format(**base_format_dict, **{'file_type': 'fasta'})
+    tax_url = base_url.format(**base_format_dict, **{'file_type': 'taxon'})
+
+    return fasta_url, tax_url
+
+
+def _retrieve_data_from_midori2(fasta_url, tax_url):
+
+    print('\nDownloading and processing raw files ... \n')
+
+    with tempfile.TemporaryDirectory() as tmpdirname:
+        for url in [fasta_url, tax_url]:
+            compressed_fn = Path(url).name
+            uncompressed_fn = Path(url).stem
+            in_path = os.path.join(tmpdirname, compressed_fn)
+            out_path = os.path.join(tmpdirname, uncompressed_fn)
+
+            try:
+                print('Retrieving data from {0}'.format(url))
+                urlretrieve(url, in_path)
+            except Exception as e:
+                raise ValueError(
+                    f'''Unable to retrieve the following file
+                    from MIDORI2:\n{url}\n: {e}''') from e
```

Comment:
```suggestion
                    f'Unable to retrieve the following file from MIDORI2:\n{url}\n{e}'
                ) from e
```
The trippel quotes introduce an unwanted new line.